server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -399,46 +398,88 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
             }
+            final var survivingIndices = updatedRepositoryData.getIndices();
             deleteIndices(
                 Optional.ofNullable(finalSnapshotInfo)
-                    .map(info -> info.indices().stream().map(repositoryData::resolveIndexId).collect(Collectors.toList()))
+                    .map(info -> info.indices().stream().filter(survivingIndices::containsKey)
+                        .map(updatedRepositoryData::resolveIndexId).collect(Collectors.toList()))
                     .orElse(Collections.emptyList()),
                 snapshotId,
                 ActionListener.map(listener, v -> {
-                    try {
-                        blobStore().blobContainer(basePath().add("indices")).deleteBlobsIgnoringIfNotExists(
-                            unreferencedIndices.stream().map(IndexId::getId).collect(Collectors.toList()));
-                    } catch (IOException e) {
-                        logger.warn(() ->
-                            new ParameterizedMessage(
-                                "[{}] indices {} are no longer part of any snapshots in the repository, " +
-                                    "but failed to clean up their index folders.", metadata.name(), unreferencedIndices), e);
-                    }
+                    cleanupStaleIndices(indicesBlobContainer, foundIndices, survivingIndices);
                     return null;
                 })
             );
         }
     }
 
+    private void cleanupStaleIndices(BlobContainer indicesBlobContainer, Map<String, BlobContainer> foundIndices,
+                                     Map<String, IndexId> survivingIndices) {
+        try {
+            final Set<String> survivingIndexIds = survivingIndices.values().stream()
+                .map(IndexId::getId).collect(Collectors.toSet());
+            final List<String> toDelete = new ArrayList<>();
+            for (Map.Entry<String, BlobContainer> indexEntry : foundIndices.entrySet()) {
+                final String indexSnId = indexEntry.getKey();
+                try {
+                    if (survivingIndexIds.contains(indexSnId) == false) {
+                        logger.debug("[{}] Found stale index [{}]. Cleaning it up.", metadata.name(), indexSnId);
+                        deleteContents(indexEntry.getValue());
+                        toDelete.add(indexSnId);
+                        logger.debug("[{}] Cleaned up stale index [{}].", metadata.name(), indexSnId);
+                    }
+                } catch (IOException e) {
+                    logger.warn(() -> new ParameterizedMessage(
+                        "[{}] index {} is no longer part of any snapshots in the repository, " +
+                            "but failed to clean up their index folders.", metadata.name(), indexSnId), e);
+                }
+            }
+            try {
+                indicesBlobContainer.deleteBlobsIgnoringIfNotExists(toDelete);
```

Comment:
this is still there to clean-up directories, right? We're somewhat breaking abstractions here.
I wonder if we should provide a `deleteDirectoryBlob` method on BlobContainer that can be more efficiently be implemented by `FSBlobContainer` and `HDFSBlobContainer`, instead of now iterating over all files for those as well. Alternatively, we could have a `deleteAll` method on BlobContainer that completely removes the BlobContainer, including the dir that it's part of.

Reply:
It would also make `deleteContents`, which is further below, an implementation detail.

Reply:
As far as the abstractions go: I'm wondering if that is still true in the real world. I just tested this and if I create a few objects in a folder and delete them, I still get the common prefix listed until I delete it (even in the S3 UI I can create an empty folder that then comes up as a common prefix).
Also S3 nowadays treats any object ending in `/` as a folder and doesn't record its size or metadata (see https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html). The same goes for Azure and GCS afaik.

That said, I think it's technically fine to make this logic an implementation detail of `BlobContainer`, I mainly liked the current approach for its simplicity in back porting :)
Practically, I'm also not sure we get much out of it (yet):
* For FS and HDFS a recursive/folder delete will still have to go through all the directories and list all the files and the delete one by one. `FileTreeWalker` might be a little more efficient here algorithmically but I don't think it practically matters for real world NFS latencies?
* For Azure + S3 implementations I could see a stronger case because we could massively optimize here by moving to non-blocking IO SDK releases and run listing and deleting in parallel (technically we could do this already in Azure where we now have the internal thread pool ...). Since we need the listing functionality publicly exposed anyway for the cleanup algorithm, maybe we could do that refactoring just in `7.x` and `8` and look into possible optimizations there and keep changes for `6.x` as short as possible?

Reply:
Sorry forgot to add: having the container delete itself is a little tricky in the current implementation(s) of the container and would require more changes than this approach because the container doesn't always have "all the information" available to delete itself (it's all trivial changes but not sure we want that noise for 6.x when it's just making the code look a little nicer there?).

Reply:
>  if I create a few objects in a folder and delete them, I still get the common prefix listed until I delete it

perhaps we should test for this in AbstractThirdPartyRepositoryTestCase to check that all implementations guarantee this behavior

Reply:
> For FS and HDFS a recursive/folder delete will still have to go through all the directories and list all the files and the delete one by one.

I'm not sure about file systems in general, but for HDFS at least, doing a recursive delete on the folder results in a single RPC call to the namenode, whereas the file-by-file delete does many many sequential RPC calls, each with significant latency.

I think we should provide folder deletion methods at the BlobContainer level so that each  type can have an efficient implementation.

Let's create a separate predecessor PR for that.

Reply:
Right .. I underestimated HDFS there, my bad -> on it with the predecessor PR.

> perhaps we should test for this in AbstractThirdPartyRepositoryTestCase to check that all implementations guarantee this behavior

Sounds good, I'll add that test in the next PR :)

Reply:
I opened https://github.com/elastic/elasticsearch/pull/43281 now that has custom (and optimized) delete calls for all the implementations individually :)